### PR TITLE
Reuse custom-install-script flag for generic userdata

### PR DESCRIPTION
**Pull request by jakefhyde**
https://github.com/rancher/rancher/issues/41023

Currently, v2prov only sent a simple script to be used for cloud-init, but as more complex use cases have arisen the need to support a fully rendered cloud-config has surfaced. Since there is no standardization across machine providers in regards to a user-data flag, the `custom-install-script` flag has been repurposed to work with either. Additionally, support has been added for jinja templates in cloud-config, which is a requirement for CAPI.

### CODE REVIEW

**Review by a-blender — APPROVED** [review]
Lgtm. I'm curious though, what happens in v2prov if a machine-driver specific user data flag and generic user data flag are passed at the same time?

**Comment by jakefhyde** [comment]
> Lgtm. I'm curious though, what happens in v2prov if a machine-driver specific user data flag and generic user data flag are passed at the same time?

Can't use both, in that case I toss it out. Unless @Oats87 strongly objects. Also, I feel like I will eventually have to support merging of the two fields, but I'm not opening that can of worms just yet.

File: commands/create.go [hunk 1]
```diff
@@ -524,17 +525,30 @@ func updateUserdataFile(driverOpts *rpcdriver.RPCFlags, machineName, hostname, u
...
 
-	if err := replaceUserdataFile(machineName, machineOS, hostname, userdataContent, customScriptContent, modifiedUserdataFile); err != nil {
-		return err
+	if bytes.HasPrefix(customScriptContent, []byte("#cloud-config")) || bytes.HasPrefix(customScriptContent, []byte("## template: jinja\n#cloud-config")) {
+		// The custom install script is a cloudinit file. This is specifically to address v2prov, where there is no
+		// machine-driver agnostic user-data flag.
+		if userdataFile != "" {
+			return fmt.Errorf("cannot pass both machine-driver specific user-data flag %s and generic user-data flag %s in '#cloud-config' format", userdataFlag, customInstallScriptFlag)
+		}
+		_, err = modifiedUserdataFile.Write(customScriptContent)
```
Comment (kinarashah) [inline commands/create.go:541]: If the custom script can get complex, would it benefit us to gzip encode it same as https://github.com/rancher/machine/blob/24ffad0c7eb9304db2a883664ab155e7dabb0344/commands/create.go#L650?

File: commands/create.go [hunk 2]
```diff
@@ -524,17 +525,30 @@ func updateUserdataFile(driverOpts *rpcdriver.RPCFlags, machineName, hostname, u
...
 		return fmt.Errorf("[updateUserdataFile] unable to create tempfile [%v]\nError returned\n[%v]", modifiedUserdataFile, err)
 	}
 	defer modifiedUserdataFile.Close()
 
-	if err := replaceUserdataFile(machineName, machineOS, hostname, userdataContent, customScriptContent, modifiedUserdataFile); err != nil {
-		return err
+	if bytes.HasPrefix(customScriptContent, []byte("#cloud-config")) || bytes.HasPrefix(customScriptContent, []byte("## template: jinja\n#cloud-config")) {
+		// The custom install script is a cloudinit file. This is specifically to address v2prov, where there is no
+		// machine-driver agnostic user-data flag.
+		if userdataFile != "" {
```
Comment (jiaqiluo) [inline commands/create.go:538]: Is it better to check the following instead? 
```suggestion
		if bytes.HasPrefix(userdataContent, []byte("#cloud-config")) {
```

**Comment by jakefhyde** [comment]
Closing as we are bumping to Q4

**Pull request closed without merging**